docs/proposals/20190610-machine-states-preboot-bootstrapping.md

```diff
@@ -0,0 +1,436 @@
+        - Type: `ControllerPhaseType`
+        - Description: The state of the bootstrap provider.
+    - **Infrastructure [optional]**
+        - Type: `ControllerPhaseType`
+        - Description: The state of the infrastructure provider.
+
+```go
+type Bootstrap struct
+```
+- **To add**
+    - **ConfigRef [optional]**
+        - Type: `*corev1.TypedLocalObjectReference`
+        - Description: ConfigRef is a reference to a bootstrap provider-specific resource that holds configuration details. The reference is optional to allow users/operators to specify Bootstrap.Data without the need of a controller.
+    - **Data [optional]**
+        - Type: `*string`
+        - Description: Data contains the bootstrap data, such as cloud-init details scripts. If nil, the Machine should remain in the Pending state.
+
+## States and transitions
+The Cluster API Machine Controller is responsible for managing the lifecycle of a Machine, including its state transitions. This is the only controller that should be writing updates to Machines.
+
+---
+### Pending
+```go
+// MachinePhasePending is the first state a Machine is assigned by
+// Cluster API Machine controller after being created.
+MachinePhasePending = MachinePhaseType("pending")
+```
+
+#### Transition Conditions
+- `Machine.Status.Phase` is empty string
+
+#### Expectations
+- When `Machine.Spec.Bootstrap.Data` is:
+    - `<nil>`, expect the field to be set by an external controller.
+    - `""` (empty string), expect the bootstrap step to be ignored.
+    - `"..."` (populated by user or from the bootstrap provider), expect the contents to be used by a bootstrap (pre-Running transition) or infrastructure (Provisioning state) provider.
+
+---
+### Provisioning
+```go
+// MachinePhaseProvisioning is the state when the
+// Machine infrastructure is being created.
+MachinePhaseProvisioning = MachinePhaseType("provisioning")
+```
+
+#### Transition Conditions
+- `Machine.Spec.Bootstrap.ConfigRef`->`Status.Phase` is “Ready”
+- `Machine.Spec.Bootstrap.Data` is not `<nil>`
+
+#### Expectations
+- Machine’s infrastructure to be in the process of being provisioned.
+
+---
+### Provisioned
+```go
+// MachinePhaseProvisioned is the state when its
+// infrastructure has been created and configured.
+MachinePhaseProvisioned = MachinePhaseType("provisioned")
+```
+
+#### Transition Conditions
+- `Machine.Spec.InfrastructureRef`->`Status.Phase` is “Ready”
+- `Machine.Spec.ProviderID` is not `<nil>`
+
+#### Expectations
+- Machine’s infrastructure has been created and the compute resource is available to be configured.
+
+---
+### Running
+```go
+// MachinePhaseRunning is the Machine state when it has
+// become a Kubernetes Node in a Ready state.
+MachinePhaseRunning = MachinePhaseType("running")
+```
+
+#### Transition Conditions
+- A Kubernetes Node is found with the same `Machine.Spec.ProviderID` and its state is “Ready”
+
+#### Expectations
+- Machine controller should set `Machine.Status.NodeRef`.
```

Comment:
Isn't this the responsibility of the node controller rather than the machine controller?

Reply:
You're that it currently is: https://github.com/kubernetes-sigs/cluster-api/blob/86f4e2c49ab9cbc377f12a2e6252bb3f9a490e90/pkg/controller/node/node.go#L45

Can we clarify this @vincepri?

Reply:
The Node controller, as it is, would be removed in `v1alpha2`. As far as I can tell, providers aren't using it and it simply doesn't work when the controllers are running in a management cluster.